contracts/upgradeable_contracts/erc20_to_native/HomeBridgeErcToNative.sol

```diff
@@ -183,25 +184,23 @@ contract HomeBridgeErcToNative is
         uint256 valueToMint = _shiftValue(_value);
         address feeManager = feeManagerContract();
         if (feeManager != address(0)) {
-            uint256 fee = calculateFee(valueToMint, false, feeManager, FOREIGN_FEE);
+            uint256 fee = calculateFee(valueToMint, feeManager, FOREIGN_FEE);
             distributeFeeFromAffirmation(fee, feeManager, _txHash);
             valueToMint = valueToMint.sub(fee);
         }
         blockReward.addExtraReceiver(valueToMint, _recipient);
         return true;
     }
 
+    /**
+     * @dev Internal function to be called when enough signatures are collected.
+     * Distributed the fee for collecting signatures.
+     * @param _message encoded message signed by the validators.
+     */
     function onSignaturesCollected(bytes _message) internal {
-        address feeManager = feeManagerContract();
-        if (feeManager != address(0)) {
-            address recipient;
-            uint256 amount;
-            bytes32 txHash;
-            address contractAddress;
-            (recipient, amount, txHash, contractAddress) = Message.parseMessage(_message);
-            uint256 fee = calculateFee(amount, true, feeManager, HOME_FEE);
-            distributeFeeFromSignatures(fee, feeManager, txHash);
-        }
+        (address recipient, uint256 amount, bytes32 txHash, ) = Message.parseMessage(_message);
+        (address feeManager, uint256 fee) = _restoreCalculatedFee(recipient, amount);
+        distributeFeeFromSignatures(fee, feeManager, txHash);
```

Comment:
It looks like this approach will not fix the original issue.
Imagine the situation when the finalization threshold on the bridge is 20, and there is a service which sends some reward through the bridge every 10 blocks.

Pre-req: the fee is `f1` for the block `X-1`
Step 1: the block `X`, the service sends the reward `5` to Alice, the `f1*5` tokens are not burnt.
Step 2: the block `X+5`, the bridge governance changes the fee to `f2`
Step 3: the block `X+10`, the service sends the reward `5` to Alice and stops its operations, the `f2*5` tokens are not burnt.
Step 4: the block `X+20`, the oracles send their signatures and `f2*5` tokens are distributed among the reward recipients
Step 5: the block `X+30`, the oracles send their signatures and if `f2` < `f1`, `f2*5` tokens are distributed among the reward recipients but `(f1-f2)*5` tokens will remain on the bridge contract balance. If `f2` > `f1`, the reward distribution will fails since the balance of the bridge contract is less than required.

Reply:
Actually, not exactly, in the current implementation, the second transfer cannot be submitted and processed until the signatures for the first transfer are collected. 

At https://github.com/poanetwork/tokenbridge-contracts/blob/62cab6e866e382b9f87f8945c93d9e1be6cadbf4/contracts/upgradeable_contracts/RewardableBridge.sol#L144-L151 I ensure that there are no any pending fees for the pair `(user, amount)`. Until this pending fee will be popped out from storage by https://github.com/poanetwork/tokenbridge-contracts/blob/62cab6e866e382b9f87f8945c93d9e1be6cadbf4/contracts/upgradeable_contracts/RewardableBridge.sol#L159-L166, user won't be able to submit exactly the same transfer.

I think it is possible to extend this functionality further, so that we can avoid such blocking behaviour, but for that we will need to maintain the stack/queue of pending transfer fees for a particular pair of `(user, amount)`. It will increase transaction gas usage, but since such logic is located on the home side of the bridge, it should not be a big problem.

What do you think of a such approach?

Reply:
> the second transfer cannot be submitted and processed until the signatures for the first transfer are collected.

OK. I missed this part. I believe it will be hard for a service developer to explain such code behavior. Especially it is risky for the cases when the oracles are experiencing the issue with the RPC connectivity and that is why they do not provide signatures for few minutes. It could break the logic of that 3rd party service.

Reply:
> I think it is possible to extend this functionality further, so that we can avoid such blocking behaviour, but for that we will need to maintain the stack/queue of pending transfer fees for a particular pair of (user, amount). It will increase transaction gas usage, but since such logic is located on the home side of the bridge, it should not be a big problem.

Based on the previous experience it seems that it is cheaper to assume that we can manage the stop of the bridge operations for such kind of maintenance actions. The situation to change the fee value or the fee manager are quite rare so don't think it makes sense to increase the cost of transactions for the end user just because we want to make sure that the imbalance will not happen during the maintenance period that could happen once or twice during the bridge lifetime.

Reply:
Ok, I added a comment with a recommendation of stopping the bridge operations before updating fee manager contracts / home fee percentage.
Does it also make sense to add a `require` statements to ensure such behaviour? So that we do not allow to change home -> foreign fee configuration until the bridge is stopped?

Reply:
> Does it also make sense to add a require statements to ensure such behaviour?

No, the comment is enough. Will not increase the code delta without a real necessity.